src/we_random.c

```diff
@@ -312,19 +246,119 @@ static int we_rand_add(const void *buf, int num, double entropy)
 #endif
 }
 
+#ifndef WE_STATIC_WOLFSSL
+/**
+ * Generate weak entropy. Used by we_rand_bytes to add entropy for RNG.
+ * Uses thread ID, a timer value, and PID.
+ *
+ * @param  out  [out]  Buffer to fill with weak entropy. If not allocated by
+ *                     caller, this function will allocate it.
+ * @returns length of out buffer on success, 0 on failure.
+ */
+static int we_rand_weak_entropy(unsigned char** out)
+{
+    int ret = 1;
+    unsigned char* idx;
+#if OPENSSL_VERSION_NUMBER >= 0x1010000fL
+    CRYPTO_THREAD_ID threadId;
+#else
+    CRYPTO_THREADID threadId;
+    unsigned long threadIdHash;
+#endif /* OPENSSL_VERSION_NUMBER >= 0x1010000fL */
+#ifndef _WIN32
+    struct timeval tv;
+    unsigned long timer;
+    pid_t pid;
+#else
+    LARGE_INTEGER timer;
+    DWORD pid;
+#endif /* !_WIN32 */
+
+#if OPENSSL_VERSION_NUMBER >= 0x1010000fL
+    threadId = CRYPTO_THREAD_get_current_id();
+#else
+    CRYPTO_THREADID_current(&threadId);
+    threadIdHash = CRYPTO_THREADID_hash(&threadId);
+#endif /* OPENSSL_VERSION_NUMBER >= 0x1010000fL */
+
+    WOLFENGINE_ENTER(WE_LOG_RNG, "we_rand_weak_entropy");
+    WOLFENGINE_MSG_VERBOSE(WE_LOG_RNG, "ARGS [out = %p]", out);
+
+    if (out == NULL) {
+        WOLFENGINE_ERROR_MSG(WE_LOG_RNG, "Bad argument.");
+        ret = 0;
+    }
+
+    if (ret == 1) {
+    #ifndef _WIN32
+        if (gettimeofday(&tv, NULL) != 0) {
+            WOLFENGINE_ERROR_MSG(WE_LOG_RNG, "gettimeofday for additional "
+                "entropy failed");
+            ret = 0;
+        }
+        else {
+            timer = 1000000 * tv.tv_sec + tv.tv_usec; /* time in us */
+            pid = getpid();
+        }
+    #else
+        ret = QueryPerformanceCounter(&timer);
+        if (ret == 0) {
+            WOLFENGINE_ERROR_MSG(WE_LOG_RNG, "QueryPerformanceCounter for "
+                "additional entropy failed");
+        }
+        else {
+            pid = GetCurrentProcessId();
+        }
+    #endif /* !_WIN32 */
+    }
+
+    if (ret == 1 && *out == NULL) {
+        *out = (unsigned char*)OPENSSL_malloc(sizeof(threadId) + sizeof(timer) +
```

Comment:
Can't see a need to malloc - a buffer on the stack should be fine.

Reply:
If we do a buffer on the stack, it will be popped when this function returns, and the calling function (`we_rand_bytes`) won't be able to access it. I could make this function take a regular `unsigned char*` for the out parameter, but then `we_rand_bytes` needs to know how to size it. I can do that, it just seems less clean.